src/main/java/com/lorepo/icplayer/client/model/addon/AddonDescriptorFactory.java

```diff
@@ -190,14 +191,22 @@ public Collection<AddonEntry> getEntries() {
 	}
 	
 	public void setWirisEnabled(boolean isEnabled) {
-		if (isEnabled) {
-			if (!addonList.containsKey("MathText")) {
-				addDescriptor("MathText", "activities_menu");
+		setOptionalAddonEnabled(isEnabled, "MathText", "activities_menu");
+	}
+
+	public void setSpeechaceEnabled(boolean isEnabled) {
+        setOptionalAddonEnabled(isEnabled, "Speechace", "activities_menu");
+	}
+
+	private void setOptionalAddonEnabled(boolean isEnabled, String addonName, String category) {
+	    if (isEnabled) {
```

Comment:
usunąłbym zagnieżdżenia:
if (isEnabled && !addonList.containsKey(addonName)) {
addDescriptor(addonName, category);
} else if (addonList.containsKey(addonName)) {
addonList.remove(addonName);
}

Reply:
To chyba nie jest tożsame.
Taki przypadek:
Chcemy żeby był, czyli isEnabled = 1 ale już aktualnie jest na liście czyli contains = 1.
Pierwszy warunek nie wchodzi bo 1 && 0 (negacja warunku contains) daje nam 0 (False).
Przechodzimy do drugiego warunku i sprawdzamy czy zawiera, a wiemy, że tak czyli contains = 1, warunek spełniony.
Tak więc mając włączony addon i aktualnie go na liście i tak go usuwamy.